Authentication/LogIn/ViewModel/LoginViewModel.swift

```diff
@@ -0,0 +1,140 @@
+//
+//  LoginViewModel.swift
+//  Authentication
+//
+//  Created by Daniela Riesgo on 3/7/16.
+//  Copyright © 2016 Wolox. All rights reserved.
+//
+
+import Foundation
+import ReactiveCocoa
+import enum Result.NoError
+
+/**
+    Protocol for login view models.
+*/
+public protocol LoginViewModelType {
+    
+    var email: MutableProperty<String> { get }
+    var emailValidationErrors: AnyProperty<[String]> { get }
+    
+    var password: MutableProperty<String> { get }
+    var passwordValidationErrors: AnyProperty<[String]> { get }
+    var showPassword: MutableProperty<Bool> { get }
+    
+    var togglePasswordVisibility: Action<AnyObject?, Bool, NoError> { get }
+    var logInCocoaAction: CocoaAction { get }
+    var logInErrors: Signal<SessionServiceError, NoError> { get }
+    var logInExecuting: Signal<Bool, NoError> { get }
+    
+    var emailText: String { get }
+    var passwordText: String { get }
+    var emailPlaceholderText: String { get }
+    var passwordPlaceholderText: String { get }
+    var loginButtonTitle: String { get }
+    var registerButtonTitle: String { get }
+    var termsAndServicesButtonTitle: String { get }
+    var passwordVisibilityButtonTitle: String { get }
+    
+}
+
+/**
+    Default LoginViewModel responsible for validating entries to email and password fields 
+    and therefore enabling log in action, managing password visibility, reporting log in events
+    and communicating with the session service for executing the log in.
+*/
+public final class LoginViewModel<User: UserType, SessionService: SessionServiceType where SessionService.User == User> : LoginViewModelType {
+    
+    private let _credentialsAreValid: AnyProperty<Bool>
+    private let _sessionService: SessionService
+    
+    public let email = MutableProperty("")
+    public let emailValidationErrors: AnyProperty<[String]>
+    
+    public let password = MutableProperty("")
+    public let passwordValidationErrors: AnyProperty<[String]>
+    public let showPassword = MutableProperty(false)
+    
+    private lazy var _logIn: Action<AnyObject, User, SessionServiceError> = {
+        return Action(enabledIf: self._credentialsAreValid) { [unowned self] _ in
+            let email = Email(raw: self.email.value)!
```

Comment:
I know this won't happen but just in case, lets change this to

```swift
if let email = Email(raw: self.email.value) {
  let password = self.password.value
  self._sessionService.logIn(email, password).observeOn(UIScheduler())
} else {
  return SignalProducer(error: .InvalidCredentials(.None))
}
```